Context extension

gabrielschulhof opened the pull request:
Move the arg extension into a static library of its own and turn it off by default.

Call user context deinit before context deinit, because the user context deinit needs a still-functioning context.

Add the context extension and turn it off by default.

zherczeg commented:
This patch seems good. However I don't everybody will happy about this change:

jerry-ext/include/jerryscript-ext/arg.h → jerry-ext/arg/include/jerryscript-ext/arg.h

This will make include directories complicated.

martijnthe commented:
> jerry-ext/include/jerryscript-ext/arg.h → jerry-ext/arg/include/jerryscript-ext/arg.h

FWIW, I'm fine with it.
I don't mind having to explicitly add a header search path per util.
Having the headers separated per util It also makes it clearer for the user what headers belong to what util.

gabrielschulhof commented:
Actually, they all install to `<prefix>/include/jerryscript-ext` and are cumulative:
```
file(GLOB INCLUDE_FILES_ARG "${INCLUDE_ARG}/jerryscript-ext/*")
install(FILES ${INCLUDE_FILES_ARG} DESTINATION include/jerryscript-ext/)
...
file(GLOB INCLUDE_FILES_CONTEXT "${INCLUDE_CONTEXT}/jerryscript-ext/*")
install(FILES ${INCLUDE_FILES_CONTEXT} DESTINATION include/jerryscript-ext/)
```
So, only the tests need the many include paths.

akosthekiss commented:
My first thoughts: 
- Why "Move the arg extension into a static library of its own and turn it off by default"? We have had this discussion before and concluded to have a single library for jerry-ext, why divert from that?
- In the new build files, there are include paths to internal directories. That flashes very big warning signs to me.

gabrielschulhof commented:
@akosthekiss 

The reason I went with separate libraries is for the case where all of them are turned off. If you go with one library (`-DJERRY_EXT=ON`) but turn off everything else (`-DJERRYX_ARG=OFF`) and (`-DJERRYX_CONTEXT=OFF`) then what do you build? I mean, if we go with one library, we still want to give the option to turn parts of it off, right?

I mean, we could have
```
if(JERRYX_ARG OR JERRYX_CONTEXT OR ...)
  add_library(${JERRYX_EXT_NAME} STATIC)
  if(JERRYX_ARG)
    target_link_libraries(${JERRYX_EXT_NAME} ${JERRYX_ARG_NAME})
  endif()
  if (JERRYX_CONTEXT)
    target_link_libraries(${JERRYX_EXT_NAME} ${JERRYX_CONTEXT_NAME})
  endif()
```
at the end of `jerry-ext/CMakeLists.txt` but it's pretty awkward to build conditionally upon *n* different flags OR-ed together.

We [talked about](https://github.com/jerryscript-project/jerryscript/pull/1784#discussion_r114262708) referring to internal API from extensions, and the conclusion was that it was OK because extensions are not mere users of JerryScript, but parts of JerryScript.

gabrielschulhof commented:
OTOH, I guess we could conditionally merge the source files, and then if the list of source files is empty, not define the target. i.e.,
```
if (JERRYX_ARG)
  file(GLOB JERRYX_ARG_SOURCES arg/*.c)
  set(SOURCES_EXT ${SOURCES_EXT} ${JERRYX_ARG_SOURCES})
endif()
if (JERRYX_CONTEXT)
  file(GLOB JERRYX_CONTEXT_SOURCES arg/*.c)
  set(SOURCES_EXT ${SOURCES_EXT} ${JERRYX_CONTEXT_SOURCES})
endif()
...
if (NOT SOURCES_EXT STREQUAL "")
  add_target(JERRY_EXT_NAME STATIC ${SOURCES_EXT})
endif()
```
Would you like that better?

akosthekiss commented:
@gabrielschulhof 

> if we go with one library, we still want to give the option to turn parts of it off, right?

Do we? Why? What do we gain by disabling parts of an extension library? If some parts of it are not used, they don't contribute to the final binary of an application.

> We talked about referring to internal API from extensions, and the conclusion was that it was OK 

I'm not sure I completely agree with that (I know that it's my bad that I haven't expressed my thoughts there). If we may access internal code/API from extensions, then why not call `ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED)` instead of `jerry_create_undefined ()`, and all that? I guess that would trigger some LBTMs. Then why is accessing jmem.h and jrt.h any different?

gabrielschulhof commented:
> > if we go with one library, we still want to give the option to turn parts of it off, right?
> 
> Do we? Why? What do we gain by disabling parts of an extension library? If some parts of it are not used, they don't contribute to the final binary of an application.

If we link a bunch of object files into `jerry-ext.a` that will never be used, that wastes space, right? OTOH, if you can `target_link_libraries(${MY_BINARY_NAME} jerryx-needed1 jerryx-needed2 ...)` then you never link anything into your binary that you're not gonna use.

If I don't use `jmem.h` I have no way of allocating memory, so I can't write any of these extensions. Like, the user of the extension may have some mechanism for allocating memory, but I don't know what that mechanism is. The only mechanism I can use is `jmem_heap_alloc_block ()`.

akosthekiss commented:
> If we link a bunch of object files into jerry-ext.a that will never be used, that wastes space, right?

Disk space during development: yes. Text/data space in the application binary: I don't think so. That's what we have the linker for. It only cherry-picks those objects from a (static) library that contain symbol definitions for unresolved references in the application objects. Those "parts" (objects) of a library that are not referenced by app objects are not linked into the final binary.

>If I don't use jmem.h I have no way of allocating memory

That's true. It might be a public API deficiency. Do you need anything else from jerry-core internals?

gabrielschulhof commented:
@akosthekiss 
I'm learning about linkers all the time :) Good to know no space is likely wasted in the final binary.

I need nothing else from jerry-core internals except the memory allocation functionality.

gabrielschulhof commented:
... and I'll put things back to a single `libjerry-ext.a`.

martijnthe commented on jerry-ext/context/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,55 @@
+  /** the deinitialization function */
+  jerry_user_context_deinit_cb deinit_cb;
+  /** the index of the slot */
+  int index;
```
Do you still need the runtime assignment of the index? Do you need the index at all?

The index can be written as a constant that is known at link-time:
```(&__jerryx_context_slot_myslot - &__start_jerryx_context_slots)```

This way the `jerryx_context_slot_t` can also be `static const` instead of just `static`.

gabrielschulhof replied:
Yes, you need the index, because you need to write it into the structure so that, when the user wants to call `jerryx_context_get ()` they have an index they can pass.

Look at the [definition of `JERRYX_CONTEXT_SLOT(slot_name)`](https://github.com/jerryscript-project/jerryscript/pull/1802/files#diff-64147807284c2c558336881bab2d4d0eR53).

You're basically reaching into another compilation unit and setting the index so that this other compilation unit will later retrieve the correct index.

gabrielschulhof replied:
I guess the extension informs its various users at which index they ended up when the binary into which they were linked was created.

gabrielschulhof replied:
Wait, I see what you're saying. You still need to store the index, but if it can be initialized at link time, then we don't need the loop.

gabrielschulhof replied:
I've tried that at https://gist.github.com/gabrielschulhof/2dfe8386a4a4924e9582bba1cb7f3123#file-section-picture-c-L23

and I get the error
```
section-picture.c:23:22: error: initializer element is not computable at load time
 static int x_index = (&x - __start_struct_array);
```

martijnthe replied:
Hrm... sorry my bad..

This does work though:
https://gist.github.com/martijnthe/0adeb2bd71e592544dde63bb002df550/revisions

gabrielschulhof commented:
@akosthekiss The context extension is now added the same way as arg.

@martijnthe I've removed the index. It need not be stored if `jerryx_context_get ()` receives a pointer to the static global structure defining the context to be retrieved, because it can compute the index from that pointer.

akosthekiss commented on jerry-ext/context/context.c:
```diff
@@ -0,0 +1,94 @@
+#include "section-utils.h"
+
+static int jerryx_context_slot_count = -1;
+DECLARE_SECTION (jerryx_ctx_slots, jerryx_context_slot_t)
```
(minor, consistency) why is this jerryx_ctx_slots, not jerryx_context_slots ?

gabrielschulhof replied:
Because section names cannot exceed 16 characters on __MACH__

akosthekiss replied:
Please, add a note/comment to explain the deviation

gabrielschulhof replied:
/o

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,67 @@
+#else /* !__MACH__ */
+#define DECLARE_SECTION(name, type)                   \
+extern type __start_ ## name[] __attribute__((weak)); \
+extern type __stop_ ## name[] __attribute__((weak));
```
(minor) any reason why these are not pointers but arrays? (I know, arrays and pointers are essentially the same in C, but having an array for "stop", which just marks the end of an actual array is strange.)

akosthekiss replied:
not adding the trailing semicolon here can help the C-ish use of this macro, too

gabrielschulhof replied:
Actually, having the semicolon is essential, and it's essential that there not be a semicolon in its usage because, if we ever need to implement this as a constructor, then there can't be a semicolon at the end of a function. So, if we take out the trailing semicolon and we need to re-implement this as a constructor, then all the users will have to change their code too, to remove the semicolon they added.

For example, if we go with constructors instead of sections because some platform doesn't support sections we'd have to define this as `#define DECLARE_SECTION(name, type) /* nothing */` and the user code would then have a stray semicolon which is not pedantic.

Basically it's safer to not use it with a semicolon, especially at the global level.

gabrielschulhof replied:
Even for internal usage like this, it's still a good idea, because if we need to
```C
#ifdef PLATFORM_SUPPORTS_SECTIONS
#define DECLARE_SECTION(name, type) /* leaves semicolon to callers */
#else /* !PLATFORM_SUPPORTS_SECTIONS */
#define DECLARE_SECTION(name, type) \
  void something(void) __attribute__((constructor)) {}
#endif /* PLATFORM_SUPPORTS_SECTIONS */
```
Then it's best to not also have to add an ifdef to the usage like so:
```
#ifdef PLATFORM_SUPPORTS_SECTIONS
DECLARE_SECTION(jerryx_ctx_slots, jerryx_context_slot_t);
#else
DECLARE_SECTION(jerryx_ctx_slots, jerryx_context_slot_t)
#endif
```

gabrielschulhof replied:
@akosthekiss I'm not sure where github put your comment, but yes, the `NODE_MODULE` macro is supposed to be used without a trailing semicolon:

https://nodejs.org/api/addons.html#addons_hello_world

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,58 @@
+ */
+void jerryx_context_deinit (void *user_context_p);
+
+typedef struct
```
doc needed for the public type

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,58 @@
+{                                                                                 \
+  .init_cb = (init_cb_ptr),                                                       \
+  .deinit_cb = (deinit_cb_ptr)                                                    \
+};
```
Don't add trailing semicolon. Then the user is forced to write `JERRYX_CONTEXT_DEFINE_SLOT (..., ..., ...);` (note the semicolon here), which is more C-ish that without the semicolon.

gabrielschulhof replied:
I know it seems unnatural for there to be no semicolon after what looks like a statement, but again, if we need to implement this using a constructor, then that's a function, and no semicolon may follow the closing brace of a function.

Look at https://github.com/nodejs/node/blob/master/src/node.h#L456-L473. The macro there must not have a semicolon after it when used because it's implemented as a constructor. The reason it's implemented as a constructor is because Node modules are dlopen()ed so they're not in any section at boot time.

If we (sooner or) later wish to support dynamic loading, we should make it possible with a simple recompile, rather than code change + recompile.

akosthekiss replied:
hm, you sure about this? the linked code _does_ make use of semicolons. `NODE_MODULE_X` is used in the definition of `NODE_MODULE` (same file, line 494), and if you search for `NODE_MODULE` in the repository, you can see it used with a trailing semicolon (https://github.com/nodejs/node/blob/db1087c9757c31a82c50a1eba368d8cba95b57d0/benchmark/misc/function_call/binding.cc)

gabrielschulhof replied:
Tisk tisk! Going against its own documentation: https://nodejs.org/api/addons.html#addons_hello_world

akosthekiss replied:
As it seems, more often than not (semicolons are winning by 12 to 4) :)

gabrielschulhof replied:
Congratulations to us for using `-pedantic`! That code wouldn't fly in this project *because* of the extra semicolons.

gabrielschulhof replied:
I'm not surprised though - the compulsion to follow ')' with ';' is very strong.

akosthekiss replied:
all right. let's keep the semicolons inside the defines

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,58 @@
+#define JERRYX_CONTEXT_SECTION "jerryx_ctx_slots"
+#endif /* __MACH__ */
+
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)         \
```
add doc to public macro

akosthekiss commented on docs/09.EXT-REFERENCE-CONTEXT.md:
```diff
@@ -0,0 +1,144 @@
+
+**Summary**
+
+Declares a new slot and stores it in the binary's data section named "jerryx_context_slots".
```
(note: here, the name of the section is "right", not jerryx_ctx_slots)

gabrielschulhof replied:
Doc went out of sync - sorry!

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,58 @@
+
+void *jerryx_context_get (jerryx_context_slot_t *slot_p);
+
+#define JERRYX_CONTEXT_SLOT(slot_name) \
```
add doc

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,67 @@
+ * Usage example:
+ * int index;
+ * widget_t *current_widget_p;
+ * ITERATE_OVER_SECTION ("widget_array", index, current_widget_p)
```
either the doc is incorrect, or the use. right now, only `for ITERATE_OVER_SECTION` works.

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,67 @@
+ * extern type __top_ ## name[] __attribute__((weak));
+ *
+ * The macro is to be succeeded by the body of a for-loop. Inside the body of the loop, the macro sets @index and
+ ** @item_p as follows:
```
double asterisk not needed

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,67 @@
+ * @item_p will point to the current item.
+ *
+ * Usage example:
+ * int index;
```
you can wrap this in
```
@code
c example comes here
@endcode
```
(helps doxygen)

gabrielschulhof replied:
Awesome! Thanks!

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Iterate over items in an array stored in a section
+ *
+ * @name is the name of the section.
```
AFAIK this is not proper doxygen syntax. You can describe function/macro parameters by
```
@param name description till the end of the line
```
and you can reference them anywhere in the text by `bla bla @p name bla bla`. (So, not just simply prefixing the name with `@`)

gabrielschulhof replied:
Oh, right. Sorry! Too many years writing gtk-doc :)

gabrielschulhof commented on jerry-ext/CMakeLists.txt:
```diff
@@ -20,11 +20,12 @@ project (${JERRY_EXT_NAME} C)
 add_library(${JERRY_EXT_NAME} STATIC ${SOURCE_EXT})
 
 target_include_directories(${JERRY_EXT_NAME} PUBLIC ${INCLUDE_EXT})
+target_include_directories(${JERRY_EXT_NAME} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/private-headers")
```
Yes, because the module extension will use the same file.

akosthekiss replied:
Sorry, I don't understand the answer. "Yes", it is needed, or "yes", it could go into the context directory?

gabrielschulhof replied:
Sorry for being unclear! I wish for this file to remain in a directory that can be referenced from multiple subdirectories of `jerry-ext/` because both `jerry-ext/context/context.c` and the future `jerryx-ext/module/module.c` will `#include` it.

akosthekiss commented on jerry-ext/context/context.c:
```diff
@@ -0,0 +1,96 @@
+  size_t context_size;
+  jerryx_context_slot_t *slot_p = NULL;
+
+  if (unlikely (jerryx_context_slot_count < 0))
```
is this `unlikely` needed here, is this code performance critical? note that this is a jerry-core internal macro. should we ever be able to get rid of deep-linking to jmem.h, we will lose this, too

akosthekiss commented on jerry-ext/context/context.c:
```diff
@@ -0,0 +1,96 @@
+  int slot = (int) (slot_p - __start_jerryx_ctx_slots);
+  void **slots_p = (void **) jerry_get_user_context ();
+
+  if (jerryx_context_slot_count < 0 || !slots_p || slot < 0 || slot >= jerryx_context_slot_count)
```
(thinking aloud) could we have this in `#ifndef JERRY_NDEBUG`?

gabrielschulhof replied:
I'm not sure. Something bad should happen if you attempt to retrieve a slot without having initialized the extension.

akosthekiss replied:
true, but such out of bounds checks usually happen in debug builds only. this is kind of an assert here, we just don't have a nice macro for it right now. and asserts are usually disabled for release builds. classic C has `assert` in `assert.h` which is a no-op if `NDEBUG` is defined, but prints an error message + aborts otherwise. jerry-core has a similar `JERRY_NDEBUG`, which controls its asserts. (defining that macro helps code size and performance a lot, at the cost of losing the checks.)

gabrielschulhof replied:
OK. I will change it.

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,83 @@
+ */
+typedef struct
+{
+  /** the initialization function */
```
(style) If I'm not mistaken, we usually document struct fields in `type name; /**< doc */` format

gabrielschulhof replied:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/include/jerryscript-core.h#L119-L159

akosthekiss replied:
Hah, well spotted. We have 102 struct typedefs in the core code base, and some 8 of them are in this style (just checked). And one happens to be in the public API headers. No pressure then

gabrielschulhof replied:
I used that as a template.

akosthekiss commented on jerry-ext/private-headers/section-utils.h:
```diff
@@ -0,0 +1,69 @@
+ * extern type __top_ ## name[] __attribute__((weak));
+ *
+ * The macro is to be succeeded by the body of a for-loop. Inside the body of the loop, the macro sets @index and
+ * @item_p as follows:
```
still non-doxygen at-signs 2 lines above and below

gabrielschulhof replied:
I don't understand. I thought you could refer to arguments previously declared with `@param argument` as `@argument` later on in the documentation.

akosthekiss replied:
no, it's `@p name` later on (see http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdp , and note that \ and @ are interchangeable)

gabrielschulhof replied:
OK. Thanks for the pointer!

zherczeg commented on docs/09.EXT-REFERENCE-CONTEXT.md:
```diff
@@ -0,0 +1,142 @@
+# jerryx_context types
```
I would prefer a separate file for each extension. And I wouldn't mind removing this numbering, since they show creation order. I am not sure it is useful.

akosthekiss replied:
(note: I did not suggest to merge all extensions into a single doc. the question was whether to keep the numbering of all extension docs unique, or whether they should all start with 09 (but in different files). I know that it has some interworking with update-webpage.sh, but I've never used that tool.)

gabrielschulhof replied:
I would say "09" refers to extensions. Like man sections.

gabrielschulhof commented on docs/09.EXT-REFERENCE-CONTEXT.md:
```diff
@@ -0,0 +1,142 @@
+
+**See also**
+
+- [JERRYX_CONTEXT_SLOT](#JERRYX_CONTEXT_SLOT)
```
No, actually, I missed that one. Uppercase doesn't work.

martijnthe commented on jerry-ext/include-private/section-utils.h:
```diff
@@ -0,0 +1,69 @@
+#ifdef __MACH__
+#define DECLARE_SECTION(name, type)                                  \
+extern type __start_ ## name[] __asm("section$start$__DATA$" #name); \
+extern type __stop_ ## name[] __asm("section$end$__DATA$" #name);
```
Huh, first time I see this. Where did you learn about this?

EDIT: Just found this SO post: http://stackoverflow.com/questions/17669593/how-to-get-a-pointer-to-a-binary-section-in-mac-os-x

gabrielschulhof replied:
So, does that mean that, for the Jerry heap it doesn't use a section on OSX?

martijnthe replied:
Hmm how does the JerryScript heap have anything to do with this?

The JerryScript heap is just a "normal" static variable by default:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/jcontext/jcontext.h#L138
So it would go into the data section automatically I think?

gabrielschulhof replied:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/jcontext/jcontext.c#L30-L40

martijnthe replied:
Ah, I see, thanks.

> So, does that mean that, for the Jerry heap it doesn't use a section on OSX?
 I guess so. I don't think it matters much when running on OSX.
I think it's probably intended for use on tiny systems where it may matter where you put the heap (you may have different types of RAM, switch of certain regions of RAM, etc.)

gabrielschulhof replied:
Gotcha.

martijnthe commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,84 @@
+ * @param deinit_cb_ptr a pointer to a function of type jerry_user_context_deinit_cb to be called when freeing a slot.
+ */
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)         \
+static jerryx_context_slot_t __jerryx_context_slot_ ## slot_name                  \
```
`static const`?  Would probably also need to sprinkle `const` around in other places when changing it.

martijnthe commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,84 @@
+ */
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)         \
+static jerryx_context_slot_t __jerryx_context_slot_ ## slot_name                  \
+__attribute__((used, section(JERRYX_CONTEXT_SECTION), aligned(sizeof(void *)))) = \
```
Wonder if it makes sense to have a macro in `section-utils.h` for this attribute (which would also take care of adding `__DATA,` for mach-o.

gabrielschulhof replied:
The problem is that context.h is public whereas section-utils.h is private.

So, I guess I should publish section-utils.h - maybe as jerryscript-ext/section.h - because being able to easily manipulate sections seems like it might be beneficial.

Who says you can't have a header-only extension, eh? :)

martijnthe replied:
Works for me :)

gabrielschulhof commented:
> Starting to look very slick :)

@martijnthe Thanks!

akosthekiss commented on jerry-ext/include/jerryscript-ext/context.h:
```diff
@@ -0,0 +1,78 @@
+ * @param deinit_cb_ptr a pointer to a function of type jerry_user_context_deinit_cb to be called when freeing a slot.
+ */
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)                                        \
+static const jerryx_context_slot_t __jerryx_context_slot_ ## slot_name JERRYX_SECTION_APPEND(jerryx_ctx_slots) = \
```
(style) indent macro body with 2 spaces

akosthekiss commented on jerry-ext/include/jerryscript-ext/section.h:
```diff
@@ -0,0 +1,81 @@
+ * limitations under the License.
+ */
+
+#ifndef JERRY_EXT_SECTION_H
```
Note 1: Do we want to expose the contents of this header to library users as an extension on its own, or we just want to (potentially) share it between public headers of extensions? To me, it does not look like a jerryscript-related extension but rather a very low-level technicality needed by some extensions. If that's the case, we have an established precedent of naming such a file as xxx.impl.h. (Note that in that case public markdown doc is not needed, even superfluous I'd say.)

Note 2: header macro guard should be JERRYX_xxx_H (i.e., JERRYX_SECTION_H or JERRYX_SECTION_IMPL_H)

gabrielschulhof replied:
@martijnthe and I [both think](https://github.com/jerryscript-project/jerryscript/pull/1802#discussion_r115692722) it warrants publication.

akosthekiss replied:
I've seen that discussion.

> The problem is that context.h is public whereas section-utils.h is private.
> So, I guess I should publish section-utils.h - maybe as jerryscript-ext/section.h - because being able to easily manipulate sections seems like it might be beneficial.

Beneficial, but only for the purpose of implementing extensions, not for users of extensions.

> Who says you can't have a header-only extension, eh? :)
 
Consider this as kind of a doubt :)

gabrielschulhof replied:
OK. I'll add `.impl` and remove the doc.

akosthekiss commented on jerry-ext/include/jerryscript-ext/section.h:
```diff
@@ -0,0 +1,81 @@
+#define JERRYX_SECTION_NAME(name) #name
+#endif /* __MACH__ */
+
+#define JERRYX_SECTION_APPEND(name) \
```
Note 1: perhaps _ATTRIBUTE instead of _APPEND, because that's what it is
Note 2: doc missing
Note 3: indent macro body

akosthekiss commented on jerry-ext/include/jerryscript-ext/section.h:
```diff
@@ -0,0 +1,81 @@
+#ifdef __MACH__
+#define JERRYX_SECTION_DECLARE(name, type)                                 \
+extern const type __start_ ## name[] __asm("section$start$__DATA$" #name); \
+extern const type __stop_ ## name[] __asm("section$end$__DATA$" #name);
```
might have missed the response: why array instead of pointer?

gabrielschulhof replied:
I guess that's how I saw the example on the Internet. I'll change it to a pointer.

gabrielschulhof replied:
Actually, changing it to a pointer doesn't work, because the `__stop_ ## name` becomes null, and it tries to allocate a pile of memory. I guess pointers and arrays are not the same to the linker.

gabrielschulhof replied:
To illustrate:

https://gist.github.com/gabrielschulhof/02fa140c2c4f93c82ee83e7773bdeb90

akosthekiss replied:
I've learned something new today, thanks

gabrielschulhof replied:
So, I guess the semantics of `__start_`/`__stop_` are such that the value of the symbol is the actual contents of the executable at that location. So, if the symbol is declared as a pointer, then the value of the pointer is determined by the contents in the executable file at that location - which are not the intended semantics. In contrast, declaring it as an array informs *somebody* (*shrug*) that it's supposed to initialize the variable to a *pointer to* the data at that location.

gabrielschulhof replied:
*somebody* is the linker, I guess.

gabrielschulhof replied:
Glad to be of service :)

martijnthe replied:
You may be able to drop the `[]` (but not make it a pointer) and use `&__start__name` and `&__stop__name`, but I'm not sure if that makes it any clearer...

gabrielschulhof replied:
Yeah, that's the alternative, and I think semantically it's better to indicate that it is an array - because it is an array :)

akosthekiss commented on jerry-ext/CMakeLists.txt:
```diff
@@ -21,15 +21,18 @@ set(INCLUDE_EXT "${CMAKE_CURRENT_SOURCE_DIR}/include")
 add_library(${JERRY_EXT_NAME} STATIC ${SOURCE_EXT})
 
 target_include_directories(${JERRY_EXT_NAME} PUBLIC ${INCLUDE_EXT})
+target_include_directories(${JERRY_EXT_NAME} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/include-private")
```
is this still needed?

akosthekiss commented on jerry-ext/context/context.c:
```diff
@@ -0,0 +1,98 @@
+
+#include <string.h>
+#include "jerryscript-ext/context.h"
+#include "jerryscript-ext/section.h"
```
isn't this already included via the public context.h header?

gabrielschulhof commented on docs/09.EXT-REFERENCE-CONTEXT.md:
```diff
@@ -0,0 +1,139 @@
+# jerryx_context types
```
It is brief, though, so I don't think I can say much more than `## Data types`.

akosthekiss commented on docs/12.EXT-REFERENCE-SECTION.md:
```diff
@@ -0,0 +1,72 @@
+# Helper macros
```
this doc has not been removed

gabrielschulhof replied:
Sorry! It's gone now.

gabrielschulhof commented:
@akosthekiss do you want me to squash?

akosthekiss commented:
@gabrielschulhof eventually, that will be a must

gabrielschulhof commented:
@akosthekiss I squashed the commits, and updated the commit message for the extension to reflect the changed name of the section ("jerryx_*ctx*_slots") and the fact that the index is no longer stored in the slot.

I also left the commit affecting core separate, because it touches a different part of the repo. Nevertheless, I can squash that in too, if you'd like.

akosthekiss commented:
@gabrielschulhof PRs are merged in a "one commit per PR" style. so, if you'd like to keep the change to the core separate, you'll need a separate PR for that, too (although, I don't think that this minor change deserves that)

gabrielschulhof commented:
@akosthekiss squash ALL the commits! :)

gabrielschulhof commented:
OK ... weird that it breaks ...

gabrielschulhof commented:
Oh, I see. There was a rename of the types.

akosthekiss commented:
still LGTM (thanks for fixing up the API reference doc)

gabrielschulhof commented:
NP - by the power of `git grep` :)

gabrielschulhof commented:
HOLD THE PRESSES!

There might be a way we can do the user context that does not require this extension at all. I'll have a PR in a bit.

gabrielschulhof commented:
https://github.com/jerryscript-project/jerryscript/pull/1815

LaszloLango commented:
@gabrielschulhof can we close this PR?

gabrielschulhof commented:
Absolutely! Sorry I left it open!
